src/aspire/abinitio/commonline_c2.py

```diff
@@ -0,0 +1,384 @@
+        :param shift_step: Resolution of shift estimation in pixels. Default = 1 pixel.
+        :param epsilon: Tolerance for the power method.
+        :param max_iter: Maximum iterations for the power method.
+        :param degree_res: Degree resolution for estimating in-plane rotations.
+        :param min_dist_cls: Minimum distance between mutual common-lines. Default = 25 degrees.
+        :param seed: Optional seed for RNG.
+        """
+        super().__init__(
+            src,
+            symmetry="C2",
+            n_rad=n_rad,
+            n_theta=n_theta,
+            max_shift=max_shift,
+            shift_step=shift_step,
+            epsilon=epsilon,
+            max_iters=max_iters,
+            degree_res=degree_res,
+            seed=seed,
+        )
+
+        self.min_dist_cls = min_dist_cls
+        self.epsilon = epsilon
+        self.max_iters = max_iters
+        self.degree_res = degree_res
+        self.seed = seed
+        self.order = 2
+
+    def _check_symmetry(self, symmetry):
+        symmetry = symmetry.upper()
+        if symmetry != "C2":
+            raise NotImplementedError(
+                f"Only C2 symmetry supported. {symmetry} was supplied."
+            )
+
+    def build_clmatrix(self):
+        """
+        Build common-lines matrix for molecules with C2 symmetry from Fourier stack of 2D images.
+        This consists of finding for each pair of images the two common-lines induced by the 2-fold symmetry.
+        """
+
+        n_img = self.n_img
+
+        if self.n_theta % 2 == 1:
+            msg = "n_theta must be even"
+            logger.error(msg)
+            raise NotImplementedError(msg)
+
+        # need to do a copy to prevent modifying self.pf for other functions.
+        pf = self.pf.copy()
+
+        # clmatrix contains the index in image i of the common line with image j
+        # for the two sets of mutual common lines.
+        clmatrix = -np.ones((2, n_img, n_img), dtype=int)
+
+        # 1D shift between common-lines.
+        shifts_1d = np.zeros((2, n_img, n_img))
+
+        # Prepare the shift phases and generate filter for common-line detection.
+        r_max = pf.shape[2]
+        shifts, shift_phases, h = self._generate_shift_phase_and_filter(
+            r_max, self.max_shift, self.shift_step
+        )
+        all_shift_phases = shift_phases.T
+        n_shifts = len(shifts)
+
+        # Apply bandpass filter, normalize each ray of each image.
+        # Note that only use half of each ray.
+        pf = self._apply_filter_and_norm("ijk, k -> ijk", pf, r_max, h)
+
+        # Pre-compute conjugated and shifted pf's.
+        pf_shifted_flipped = np.conj(pf)[:, None] * all_shift_phases[:, None]
+        pf_shifted_flipped = pf_shifted_flipped.reshape(
+            (self.n_img, n_shifts * (self.n_theta // 2), r_max)
+        )
+
+        # Search for common lines between [i, j] pairs of images.
+        # The random selection is implemented.
```

Comment:
Where is the random part?

Reply:
Oops. Good catch. I removed that part. Comment has been deleted. Thanks